File: Shared/Blog.Core/Helpers/EmailHelper.cs
namespace Blog.Core.Helpers;

using Calabonga.Microservices.Core.Extensions;
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// Email validation helper.
/// </summary>
public static class EmailHelper
{
    /// <summary>
    /// Validates emails entries.
    /// </summary>
    /// <param name="emails">emails.</param>
    /// <returns>IEnumerable.</returns>
    public static IEnumerable<string> GetValidEmails(string emails)
    {
        if (string.IsNullOrWhiteSpace(emails))
        {
            return null;
        }

        var split = emails.Split([';', '|', ' ', ','], StringSplitOptions.RemoveEmptyEntries);
        return split.Where(x => x.IsEmail());
    }
}

File: Shared/Blog.FSharp.Core/Helpers/EmailHelper.fs
namespace Blog.FSharp.Core.Helpers

open System
open System.Text.RegularExpressions

/// <summary>
/// Email validation helper.
/// </summary>
module EmailHelper =

    /// <summary>
    /// Checks if a string is a valid email.
    /// </summary>
    let private isValidEmail (email: string) : bool =
        if String.IsNullOrWhiteSpace(email) then false
        else
            let pattern = @"^[^@\s]+@[^@\s]+\.[^@\s]+$"
            Regex.IsMatch(email, pattern, RegexOptions.IgnoreCase)

    /// <summary>
    /// Validates email entries.
    /// </summary>
    /// <param name="emails">A string with emails separated by ; | , or space.</param>
    /// <returns>Sequence of valid emails.</returns>
    let getValidEmails (emails: string) : seq<string> =
        if String.IsNullOrWhiteSpace(emails) then
            Seq.empty
        else
            emails.Split([| ';'; '|'; ' '; ',' |], StringSplitOptions.RemoveEmptyEntries)
            |> Seq.map (fun x -> x.Trim())      // remove extra spaces
            |> Seq.filter isValidEmail          // use our regex validation
